docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+* `Transport`: An arbitrary mechanism to exchange bytes with a peer using IO `Stream`s across a `Connection`.
+* `Channel`: A bidirectional channel to asynchronously exchange Protobuf messages with peers addressed with `PeerID`.
+* `Router`: Maintains transport connections to relevant peers and routes channel messages.
+* Reactor: A design pattern loosely defined as "something which listens on a channel and reacts to messages".
+
+These abstractions and related concepts are illustrated in the following diagram and described in detail below.
+
+![P2P Architecture Diagram](img/adr-062-architecture.svg)
+
+### Transports
+
+Transports are arbitrary mechanisms for exchanging raw bytes with a peer. For example, a gRPC transport would connect to a peer over TCP/IP and send data using the gRPC protocol, while an in-memory transport might communicate with a peer running in another goroutine using internal byte buffers. Note that transports don't have a notion of a `peer` as such - instead, they communicate with an arbitrary endpoint address (e.g. IP address and port number), to decouple them from the rest of the P2P stack.
+
+Transports must satisfy the following requirements:
+
+* Be connection-oriented, and support both listening for inbound connections and making outbound connections using endpoint addresses.
+
+* Support multiple logical IO streams within a single connection, to take full advantage of protocols with native stream support. For example, QUIC supports multiple independent streams, while HTTP/2 and MConn multiplex logical streams onto a single TCP connection.
+
+* Provide the public key of the peer, and possibly encrypt or sign the traffic as appropriate. This should be compared with known data (e.g. the peer ID) to authenticate the peer and avoid man-in-the-middle attacks.
+
+The initial transport implementation will be a port of the current MConn protocol currently used by Tendermint, and should be backwards-compatible at the wire level. This will be followed by an in-memory transport for testing, and a QUIC transport that may eventually replace MConn.
+
+The `Transport` interface is:
+
+```go
+// Transport is an arbitrary mechanism for exchanging bytes with a peer.
+type Transport interface {
+    // Accept waits for the next inbound connection on a listening endpoint.
+    Accept(context.Context) (Connection, error)
+
+    // Dial creates an outbound connection to an endpoint.
+    Dial(context.Context, Endpoint) (Connection, error)
+
+    // Endpoints lists endpoints the transport is listening on. Any endpoint IP
+    // addresses do not need to be normalized in any way (e.g. 0.0.0.0 is
+    // valid), as they should be preprocessed before being advertised.
+    Endpoints() []Endpoint
+}
+```
+
+How the transport configures listening is transport-dependent, and not covered by the interface. This typically happens during transport construction.
```

Comment:
How many times is a transport initiated? Is there one `Transport` initiated for each transport type (e.g., one memory transport, one mconnection, etc...) when the node starts?

Reply:
That's really up to the transport, but the general pattern is to construct a transport and set it up to listen on whatever interfaces are desired, then pass it to the router. There will generally be one instance of each transport.